fv/wireguard_test.go

```diff
@@ -875,9 +906,252 @@ var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ WireGuard-Supported 3 node
 	})
 })
 
+var _ = infrastructure.DatastoreDescribe("_BPF-SAFE_ WireGuard-Supported 3-node cluster with WorkloadIPs", []apiconfig.DatastoreType{apiconfig.EtcdV3, apiconfig.Kubernetes}, func(getInfra infrastructure.InfraFactory) {
+	const nodeCount, wlPerNode = 3, 2
+
+	var (
+		infra   infrastructure.DatastoreInfra
+		felixes []*infrastructure.Felix
+		client  clientv3.Interface
+
+		// simulated host workloads
+		wlsByHost        [nodeCount][wlPerNode]*workload.Workload
+		hostNetworkedWls [nodeCount]*workload.Workload
+		externalClient   *containers.Container
+
+		cc       *connectivity.Checker
+		tcpdumps []*tcpdump.TCPDump
+	)
+
+	BeforeEach(func() {
+		// Run these tests only when the Host has Wireguard kernel module available.
+		if os.Getenv("FELIX_FV_WIREGUARD_AVAILABLE") != "true" {
+			Skip("Skipping Wireguard supported tests.")
+		}
+
+		infra = getInfra()
+		topologyOptions := wireguardTopologyOptions("WorkloadIPs", false)
+		felixes, client = infrastructure.StartNNodeTopology(nodeCount, topologyOptions, infra)
+
+		// To allow all ingress and egress, in absence of any Policy.
+		infra.AddDefaultAllow()
+
+		// initialise pods
+		for felixIdx, felixWls := range wlsByHost {
+			for wlIdx := range felixWls {
+				wlsByHost[felixIdx][wlIdx] = createWorkloadWithAssignedIP(
+					&infra,
+					&topologyOptions,
+					&client,
+					fmt.Sprintf("10.65.%d.%d", felixIdx, 2+wlIdx),
+					fmt.Sprintf("wl-f%d-%d", felixIdx, wlIdx),
+					felixes[felixIdx])
+			}
+		}
+
+		// initialise host-networked pods
+		for i, _ := range hostNetworkedWls {
+			hostNetworkedWls[i] = createHostNetworkedWorkload(fmt.Sprintf("wl-f%d-hn-0", i), felixes[i])
+		}
+
+		// initialise external client
+		externalClient = containers.Run("external-client",
+			containers.RunOpts{AutoRemove: true},
+			"--privileged", // So that we can add routes inside the container.
+			utils.Config.BusyboxImage,
+			"/bin/sh", "-c", "sleep 1000")
+		externalClient.EnsureBinary("test-connection")
+		externalClient.Exec("ip", "route", "add", wlsByHost[0][0].IP, "via", felixes[0].IP)
+
+		for i := range felixes {
+			felixes[i].TriggerDelayedStart()
+		}
+
+		// Check felix Wireguard links are ready.
+		for i := range felixes {
+			Eventually(func() string {
+				out, _ := felixes[i].ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+				return out
+			}, "10s", "100ms").Should(Not(BeEmpty()))
```

Comment:
Trying to think what could cause flakes; does `ip link show` actually return 0 bytes if there's no output?  If it returned just `\n` say then this test might fail.

Reply:
yeah it's zero bytes because `out` only contains stdout and if there is no link yet, it returns an err `Device "wireguard.cali" does not exist.` in stderr instead... 

but you do bring up a good point, maybe we can use ExecCombinedOuput instead to assert that it's not the error message instead?

Reply:
we'll leave this for now since it's also being tested this way on other tests in the same file and also in other files...